dev/src/v1/firestore_client.ts

```diff
@@ -0,0 +1,1318 @@
+ *        monotonic fashion.
+ *  *    `update_time` - The time at which a document was last updated. Changes
+ *       every time a document is modified. Does not change when a write results
+ *       in no modifications. Increases in a strict monotonic fashion.
+ *  *    `read_time` - The time at which a particular state was observed. Used
+ *       to denote a consistent snapshot of the database or the time at which a
+ *       Document was observed to not exist.
+ *  *    `commit_time` - The time at which the writes in a transaction were
+ *       committed. Any read with an equal or greater `read_time` is guaranteed
+ *       to see the effects of the transaction.
+ * @class
+ * @memberof v1
+ */
+export class FirestoreClient {
+  private _descriptors: Descriptors = {page: {}, stream: {}, longrunning: {}};
+  private _firestoreStub: Promise<{[name: string]: Function}>;
+  private _innerApiCalls: {[name: string]: Function};
+  private _terminated = false;
+  auth: gax.GoogleAuth;
+
+  /**
+   * Construct an instance of FirestoreClient.
+   *
+   * @param {object} [options] - The configuration object. See the subsequent
+   *   parameters for more details.
+   * @param {object} [options.credentials] - Credentials object.
+   * @param {string} [options.credentials.client_email]
+   * @param {string} [options.credentials.private_key]
+   * @param {string} [options.email] - Account email address. Required when
+   *     using a .pem or .p12 keyFilename.
+   * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
+   *     .p12 key downloaded from the Google Developers Console. If you provide
+   *     a path to a JSON file, the projectId option below is not necessary.
+   *     NOTE: .pem and .p12 require you to specify options.email as well.
+   * @param {number} [options.port] - The port on which to connect to
+   *     the remote host.
+   * @param {string} [options.projectId] - The project ID from the Google
+   *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
+   *     the environment variable GCLOUD_PROJECT for your project ID. If your
+   *     app is running in an environment which supports
+   *     {@link https://developers.google.com/identity/protocols/application-default-credentials Application Default Credentials},
+   *     your project ID will be detected automatically.
+   * @param {function} [options.promise] - Custom promise module to use instead
+   *     of native Promises.
+   * @param {string} [options.apiEndpoint] - The domain name of the
+   *     API remote host.
+   */
+
+  constructor(opts?: ClientOptions) {
+    // Ensure that options include the service address and port.
+    const staticMembers = this.constructor as typeof FirestoreClient;
+    const servicePath =
+      opts && opts.servicePath
+        ? opts.servicePath
+        : opts && opts.apiEndpoint
+        ? opts.apiEndpoint
+        : staticMembers.servicePath;
```

Comment:
We don't generate 3.7 code (with optional chaining and null coalescing). Even worse, we are going to rollback libraries to 3.6 because of some compatibility issues with 3.5. https://github.com/googleapis/google-auth-library-nodejs/issues/844 has some details.